enhancements/microshift/microshift-supportability-tools.md

```diff
@@ -0,0 +1,291 @@
+1. Yaml output for relevant manifests from API server.
+1. Logs for every container. These are already retrieved by CRIO plugin. May
+need easier indexing.
+1. MicroShift's configuration and version. This [enhancement](https://github.com/openshift/enhancements/pull/1304)
+might add additional information to gather: plugins, manifests, etc.
+
+Retrieving all this information involves using a kubernetes client. While
+MicroShift itself uses the go client extensively, the idea of introducing
+another command to produce this output and avoid installing `oc`/`kubectl` is
+not the best, since we would be adding runtime dependencies between a system
+component and MicroShift.
+
+### Workflow Description
+
+> _Disclaimer:_ Workflows described here illustrate manual processes.
+
+Assuming from now on MicroShift has already been installed in a R4E/RHEL based
+system and there has been some kind of issue.
+
+1. MicroShift administrator notices an issue/wants to get a system report.
+1. The administrator will log in to the failing host.
+1. The administrator will perform the debug report creation procedure:
+```bash
+sudo sos report
+```
+
+For the CI, the same procedure applies.
+
+#### Variation [optional]
+N/A
+
+### API Extensions
+N/A
+
+### Implementation Details/Notes/Constraints [optional]
+N/A
+
+### Risks and Mitigations
+There are four main risks:
+1. Upgrades and incompatible changes.
+1. `sos` release cadence.
+1. Extra required disk space.
+1. CEE having a different tool.
+
+#### Upgrades
+When upgrading a MicroShift system we may need to upgrade just MicroShift, or
+the OS as well. If it is the latter then `sos` will be updated before
+MicroShift, therefore we must ensure backwards compatibility with the plugin.
+
+If it is the former then we need to ensure newer MicroShift versions do not
+introduce incompatible changes with the reporting capabilities of the `sos`
+plugin. This may also be seen from a different perspective: the plugin should
+make no assumptions about names or locations of components whenever possible.
+In other words, the plugin must be written in the most generic way that still
+allows retrieval of relevant information.
+
+Still, if incompatible changes are introduced then we need to upgrade `sos` as
+well.
+
+#### Sos release cadence
+`sos` release cadence has been shortened to 4 weeks starting on Feb 1st 2023.
+This change has been introduced to allow faster support of new components and
+features. Since MicroShift follows the same release calendar as OpenShift,
+`sos` is faster.
+
+New versions of `sos` will simply slot in to the next regular release. If there
+is an urgent need a hotfix can be made between releases.
+
+#### Extra disk space
+Adding a plugin to `sos` and not relying on MicroShift binary means it needs
+the `oc` or `kubectl` binary installed. This package is available in the RHEL
+repos, but it comes at an extra cost of 125MB once installed.
```

Comment:
How large is the `oc` client? I'm wondering if we could add it to the `support-tools` container image that OCP uses for sos. We do however try very hard to limit the size of that image.

Reply:
Is it possible to run `sos report` from another machine? We would need to pull the kubeconfig from MicroShift host, but it would remove this requirement.

Greenboot is going to require `oc` anyway, IIRC.

Reply:
Yes, but not in the way that I think you're getting at.

`sos collect` is used to orchestrate collecting sos reports from multiple systems at once. This can be done from a workstation, so long as there is network access to a system capable of enumerating what nodes to collect from, and network access to those enumerated nodes. For example, a user that has `oc` installed and configured locally for an OCP cluster can run `sos collect` locally and it will discover the OCP nodes, connect to them (by default deploying a debug pod and leveraging it via `oc exec`, but if that fails we can SSH directly to the nodes) and collect an `sos report` from the node. **But**, the plugins/the report command itself still run directly on the nodes in question. So the `oc` collections done would still need to be executed locally on the node we're generating the report for.

Reply:
Potentially, a user could have a local `oc` configured for a microshift environment and run `sos report -o microshift` locally and that *would* get the same output from the plugin as if it were run on the microshift machine - but I don't think that's a pattern we want to rely on.

Reply:
Almost the same output, as the part of retrieving the journalctl/microshift commands wouldnt work. The journal logs can be retrieved by other plugins (in a different way, but still), but not the microshift commands.

Seems we would need full access to the node and `oc` installed.